File: interface/lib/models/address.dart
// To parse this JSON data, do
//
//     final address = addressFromJson(jsonString);

import 'dart:convert';

Address addressFromJson(String str) => Address.fromJson(json.decode(str));

String addressToJson(Address data) => json.encode(data.toJson());

class Address {
  Address({
    required this.addresses,
  });

  List<AddressElement> addresses;

  factory Address.fromJson(Map<String, dynamic> json) => Address(
        addresses: List<AddressElement>.from(
            json["addresses"].map((x) => AddressElement.fromJson(x))),
      );

  Map<String, dynamic> toJson() => {
        "addresses": List<dynamic>.from(addresses.map((x) => x.toJson())),
      };
}

class AddressElement {
  AddressElement({
    required this.addressId,
    required this.street,
    required this.city,
    required this.state,
    required this.zipCode,
    required this.detail,
    required this.latitude,
    required this.longitude,
  });

  int addressId;
  String street;
  String city;
  String state;
  String zipCode;
  String detail;
  double latitude;
  double longitude;

  factory AddressElement.fromJson(Map<String, dynamic> json) => AddressElement(
        addressId: json["AddressID"],
        street: json["street"],
        city: json["city"],
        state: json["state"],
        zipCode: json["zip_code"],
        detail: json["detail"],
        latitude: json["latitude"].toDouble(),
        longitude: json["longitude"].toDouble(),
      );

  Map<String, dynamic> toJson() => {
        "street": street,
        "city": city,
        "state": state,
        "zip_code": zipCode,
        "detail": detail,
        "latitude": latitude,
        "longitude": longitude,
      };
}

File: backend/api/handlers/user/address.go
package user

import (
	"ciceksepeti/api/handlers/errorHandler"
	"ciceksepeti/database"
	"ciceksepeti/models"

	"github.com/go-playground/validator/v10"
	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v4"
)

// AddMessage adds address to user
func AddAddress(c *fiber.Ctx) error {
	// Parse the request body and extract the address values
	var address models.Address
	if err := c.BodyParser(&address); err != nil {
		return errorHandler.BadRequest(c)
	}
	//fmt.Println("address", address)
	// Validate the address values
	err := models.Validate.Struct(address)
	if err != nil {
		for _, err := range err.(validator.ValidationErrors) {
			if err.Tag() == "required" {
				return errorHandler.RequiredField(c, err.Field())
			}
			if err.Tag() == "min" {
				return errorHandler.MinLength(c, err.Field(), err.Param())
			}
			if err.Tag() == "max" {
				return errorHandler.MaxLength(c, err.Field(), err.Param())
			}
		}
	}

	userToken := c.Locals("user").(*jwt.Token)
	claims := userToken.Claims.(jwt.MapClaims)

	userId := uint(claims["sub"].(float64))

	// Insert the address into the database
	_, err = database.DB.Exec("INSERT INTO address (userId, street, city, state, zipCode, detail, latitude, longitude) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
		userId, address.Street, address.City, address.State, address.ZipCode, address.Detail, address.Latitude, address.Longitude)
	if err != nil {
		//fmt.Println(err)
		return errorHandler.InternalServerError(c)
	}

	go database.Lof.Println("User | Address added successfully by", userId)

	// Return a success response
	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"message": "Address added successfully",
	})
}
